gdb/rust-lang.c

```diff
@@ -60,7 +70,80 @@ rust_crate_for_block (const struct block *block)
   return xstrndup (scope, len);
 }
 
-
+/* Information about the discriminant/variant of an enum */
+
+struct disr_info
+{
+  char* name; // needs to be freed
+  int field_no; // Field number in union. negative if error
+};
+
+/* Utility function to get discriminant info for a given value.  */
+
+static struct disr_info
+rust_get_disr_info (struct type *type, const gdb_byte *valaddr,
+                    int embedded_offset, CORE_ADDR address,
+                    const struct value *val) {
+
+  int i;
+  struct disr_info ret;
+  struct type *disr_type;
+  struct ui_file * temp_file;
+
+  struct value_print_options opts;
+
+  get_no_prettyformat_print_options (&opts);
+
+  ret.field_no = -1;
+
+  if (TYPE_NFIELDS (type) == 0) {
+    /* void enums should never exist as values */
+    return ret;
+  }
+
+  disr_type = TYPE_FIELD_TYPE (type, 0);
+
+  if (TYPE_NFIELDS (disr_type) == 0) {
+    /* This is a bounds check and should never
+       be hit unless Rust has changed its debuginfo
+       format */
+    return ret;
+  }
+
+  if (strcmp (TYPE_FIELD_NAME (disr_type, 0), "RUST$ENUM$DISR") != 0) {
+    /* debuginfo format has changed */
+    return ret;
+  }
+
+  temp_file = mem_fileopen ();
+
+  /* The first value of the first field (or any field)
+     is the discriminant value.  */
+  c_val_print (TYPE_FIELD_TYPE (disr_type, 0), valaddr,
+              embedded_offset + TYPE_FIELD_BITPOS (type, 0) / 8
+                              + TYPE_FIELD_BITPOS (disr_type, 0) / 8,
+              address, temp_file,
+              0, val, &opts);
+
+  ret.name = ui_file_xstrdup (temp_file, NULL);
+  ui_file_delete (temp_file);
+
+  for (i = 0; i < TYPE_NFIELDS (type); ++i) {
+
+    /* Sadly, the discriminant value paths do not match the type field
+       name paths (`core::option::Option::Some` vs `core::option::Some`)
+       However, enum variant names are unique in the last path segment and
+       the generics are not part of this path, so we can just compare those.
+       This is hackish and would be better fixed by improving rustc's
+       metadata for enums.  */
+    if(strcmp (rust_last_path_segment (ret.name),
+               rust_last_path_segment (TYPE_NAME (TYPE_FIELD_TYPE (type, i)))) == 0) {
+      ret.field_no = i;
```

Comment:
It seems like we could in theory exit the loop without setting field_no, in which case `name` would leak.